README.md
# spa-express-static-server

An opinionated static server library for Angular single page applications (SPAs).

[![Build Status](https://travis-ci.org/cgmartin/spa-express-static-server.svg?branch=master)](https://travis-ci.org/cgmartin/spa-express-static-server)
[![Dependency Status](https://david-dm.org/cgmartin/spa-express-static-server.svg)](https://david-dm.org/cgmartin/spa-express-static-server)
[![npm version](https://badge.fury.io/js/spa-express-static-server.svg)](http://badge.fury.io/js/spa-express-static-server)

## Synopsis

The spa-express-static-server is specifically meant as a reusable static file server
and Angular "HTML5 Mode" route handler. It could be replaced by Apache Web Server, Nginx, or other
static web servers. No web API/REST calls allowed here.

This static server project is meant to accompany a [SPA client](https://github.com/cgmartin/angular-spa-browserify-example)
and a set of separate Node.js microservices (REST webservice, Chat Server, Reverse Proxy).
It is designed with portability and scalability in mind (see [Twelve Factors](http://12factor.net/)).

Configuration options are passed in by the consumer or via environment variables at runtime.

## Quick Start / Usage

```bash
$ npm install spa-express-static-server --save
```

Create a `server.js` wrapper script, passing in the configuration options that apply for your app:
```js
// server.js
var staticServer = require('spa-express-static-server');
staticServer.start({
    webRootPath: './dist/web-root',
    spaBoot:     require('./spa-boot'),
    sslKeyFile:  './keys/my-domain.key',
    sslCertFile: './keys/my-domain.cert'
});
```

And run your `server.js` with optional runtime environment variables:
```bash
$ STATIC_COMPRESSION=1 STATIC_SSL=1 STATIC_PORT=443 node server.js
```

See [src/config.js](src/config.js) for a full list of the available configuration options.

### Default Environment Variables

* `STATIC_WEBROOT` : Path to the web root directory.
* `STATIC_COMPRESSION` : Enables gzip compression when set to "1".
* `STATIC_GRACEFUL_SHUTDOWN` : Wait for connections to close before stopping server when set to "1".
* `STATIC_SESSION_MAXAGE` : The time in ms until the session ID cookie should expire (default: 2 hours). This is just a tracking cookie, no session storage is used here.
* `STATIC_REV_PROXY` : The server is behind a reverse proxy when set to "1".
* `STATIC_PORT` : The port to run on (default: 8000).
* `STATIC_SSL` : Use a HTTPS server when set to "1". Enforces HTTPS by redirecting HTTP users when used with a reverse HTTP/HTTPS proxy.
* `STATIC_SSL_KEY` : Path to the SSL key file.
* `STATIC_SSL_CERT` : Path to the SSL cert file.

## Features

* **AngularJS HTML5 mode**: Catch all non-file routes and forward to index.html.
* **Security headers** using [Helmet](https://github.com/helmetjs/helmet) middleware.
* **Correlation ID Cookies**: Creates unique request and "conversation" id cookies. Useful for tracking client API requests throughout a user's session lifetime.
* **SPA Boot Configuration**: JSONP launcher that provides runtime configuration for the client.
* **Graceful shutdown**: Listens for SIGTERM/SIGINT and unhandled exceptions, and waits for open connections to complete before exiting.
* **JSON format access logs**: Great for log analysis and collectors such as Splunk, Fluentd, Graylog, Logstash, etc.
* **Enforce HTTPS**: Redirects users from HTTP urls to HTTPS.

## Contributing

1. Install [Node.js](https://nodejs.org/download/)
1. Install Gulp: `npm -g i gulp`
1. Clone this repo
1. Install dependencies: `npm i`
1. Start the app in dev mode: `npm start`
1. Point browser to <http://localhost:3000/> and watch the console for server logs

After installation, the following actions are available:

* `npm start` : Runs in development mode, starting the server and a local webserver, running linting and unit tests, and restarting upon file changes.
* `npm test` : Runs JavaScript file linting and unit tests.
* `npm run watch` : Alternative development mode - does not run servers. Only runs linting and unit tests upon file changes.

Please ensure linting and unit tests pass when submitting Pull Requests.

## Folder Structure

```
├── coverage          # Coverage reports
├── example           # Example client assets for testing
└── src
    ├── middleware    # Express middleware utilities
    ├── lib
    │   ├── app.js               # Creates and configures an express app
    │   ├── errors.js            # Custom error classes
    │   └── graceful-shutdown.js # Attempts a graceful server shutdown
    │
    ├── config.js     # Configuration options
    └── server.js     # Starts the express app on a port
```

## Libraries & Tools

The functionality has been implemented by integrating the following 3rd-party tools and libraries:

 - [Express](https://github.com/strongloop/express): Fast, minimalist web framework for node
 - [Helmet](https://github.com/helmetjs/helmet): Secure Express apps with various HTTP headers
 - [Bunyan](https://github.com/trentm/node-bunyan): A simple and fast JSON logging module for node.js services
 - [Gulp](http://gulpjs.com/): Streaming build system and task runner
 - [Node.js](http://nodejs.org/api/): JavaScript runtime environment for server-side development
 - [Mocha](http://mochajs.org/): The fun, simple, flexible JavaScript test framework
 - [Chai](http://chaijs.com/): BDD/TDD assertion library for node and the browser
 - [Sinon](http://sinonjs.org/): Standalone test spies, stubs and mocks for JavaScript
 - [Mockery](https://github.com/mfncooper/mockery): Mock Node.js module dependencies during testing

## License

[MIT License](http://cgm.mit-license.org/)  Copyright © 2015 Christopher Martin


CHANGELOG.md
0.0.12 / 2015-07-19
===================

- update: Provide a status endpoint for health checks. Enable with `config.statusRoute = '/status'`.

0.0.11 / 2015-06-21
===================

- fix: Improve graceful shutdown logic from spawning multiple timers

0.0.10 / 2015-06-13
===================

- update: Remove session tracking id

0.0.9 / 2015-06-13
==================

- update: Expose configuration option for graceful shutdown
- fix/update: Expose configuration option for gzip compression (was NODE_ENV="production")

0.0.8 / 2015-06-08
==================

- update: Added tuning options for http server (timeout and maxHeadersCount)
- update: Expose configurable compression middleware options
- fix: Session cookie maxAge default from 2 mins to 20 mins
- fix: Remove morgan node dependency

0.0.7 / 2015-06-08
==================

- update: Replaced morgan with bunyan logger

0.0.6 / 2015-06-07
==================

- fix: Relative file paths would cause res.sendFile error

0.0.5 / 2015-06-07
==================

- new: Support for custom 404/505 pages
- update: Additional commenting for internal source code, namely config options
- update: Change default session timeout from 2 hours to 20 mins

0.0.4
=====

This file was started after the release of 0.0.4.


gulpfile.js
'use strict';
var fs          = require('fs');
var path        = require('path');
var _           = require('lodash');
var gulp        = require('gulp-help')(require('gulp'));
var $           = require('gulp-load-plugins')({lazy: true});
var source      = require('vinyl-source-stream');
var buffer      = require('vinyl-buffer');
var runSequence = require('run-sequence');
var merge       = require('merge2');
var args        = require('yargs').argv;
var notifier    = require('node-notifier');
var del         = require('del');
var browserSync = require('browser-sync');
var pkg         = require('./package.json');

process.setMaxListeners(0);       // Disable max listeners for gulp

var isVerbose = args.verbose;     // Enable extra verbose logging with --verbose
var isProduction = args.prod;     // Run extra steps (minification) with production flag --prod
var isSyncEnabled = !args.nosync; // Disable browsersync with --nosync
var isWatching = false;           // Enable/disable tasks when running watch

/************************************************************************
 * Functions/Utilities
 */

// Desktop notifications of errors
function onError(err) {
    // jshint validthis: true
    notifier.notify({
        title: err.plugin + ' Error',
        message: err.message
    });
    $.util.log(err.toString());
    $.util.beep();
    if (isWatching) {
        this.emit('end');
    } else {
        process.exit(1);
    }
}

function verbosePrintFiles(taskName) {
    return $.if(isVerbose, $.print(function(filepath) {
        return taskName + ': ' + filepath;
    }))
}

/************************************************************************
 * Clean temporary folders and files
 */

gulp.task('clean-build', false, function(cb) {
    del(['.tmp', 'dist'], cb);
});

gulp.task('clean-coverage', false, function(cb) {
    del(['coverage'], cb);
});

gulp.task('clean', 'Remove all temporary files', ['clean-build', 'clean-coverage']);

/************************************************************************
 * JavaScript tasks
 */

gulp.task('lint', 'Lints all JavaScript files', function() {
    return gulp.src('src/**/*.js')
        .pipe($.plumber({errorHandler: onError}))
        .pipe(verbosePrintFiles('lint-js'))
        .pipe($.jscs())
        .pipe($.jshint())
        .pipe($.jshint.reporter('jshint-stylish', {verbose: true}))
        .pipe($.if(!isWatching, $.jshint.reporter('fail')));
});

/************************************************************************
 * Unit testing tasks
 */

function serverTestStream() {
    process.env.NODE_ENV = 'test';
    return gulp.src('src/**/*.spec.js', {read: false})
        .pipe($.plumber({errorHandler: onError}))
        .pipe($.if(args.verbose, $.print()))
        .pipe($.mocha({
            reporter: 'spec'
        }));
}

gulp.task('test-server', function() {
    return serverTestStream();
});

gulp.task('test-server-coverage', ['clean-coverage'], function(cb) {
    var coverageDir = './coverage';
    gulp.src(['src/**/*.js', '!**/*.spec.js'])
        .pipe($.if(args.verbose, $.print()))
        .pipe($.istanbul({ // Covering files
            //instrumenter: isparta.Instrumenter,
            includeUntested: true
        }))
        .pipe($.istanbul.hookRequire()) // Force `require` to return covered files
        .on('finish', function() {
            serverTestStream()
                .pipe($.istanbul.writeReports({
                    dir: coverageDir,
                    reportOpts: {dir: coverageDir},
                    reporters: ['text', 'text-summary', 'json', 'html']
                }))
                .pipe($.istanbulEnforcer({
                    thresholds: {
                        statements: 0,
                        branches: 0,
                        lines: 0,
                        functions: 0
                    },
                    coverageDirectory: coverageDir,
                    rootDirectory: ''
                }))
                .on('end', cb);
        });
});

gulp.task('test', 'Run unit tests', function(cb) {
    runSequence('clean', 'lint', 'test-server-coverage', cb);
});

/************************************************************************
 * Build / Watch / Reload tasks
 */

gulp.task('build', 'Builds the source files into a distributable package', function(cb) {
    runSequence('clean-build', 'build-iterate', cb);
}, {
    options: {
        'prod':    'Enable production minification, sourcemaps, etc.',
        'verbose': 'Display debugging information'
    }
});

gulp.task('build-iterate', false, function(cb) {
    runSequence('lint', 'test-server', cb);
});

gulp.task('watch', false, function() {
    isWatching = true;
    gulp.watch('src/**/*.js', ['build-iterate']);
});

function startBrowserSync(port) {
    if (!isSyncEnabled) {
        return;
    }
    if (browserSync.active) {
        browserSync.reload();
        return;
    }

    port = port || 8000;
    $.util.log('Starting browser-sync on port ' + port);

    var options = {
        proxy: 'localhost:' + port,
        port: 3000,
        files: ['./example/web-root/**'],
        ghostMode: {
            clicks: true,
            location: false,
            forms: true,
            scroll: true
        },
        injectChanges: true,
        logFileChanges: true,
        logLevel: 'info',
        logPrefix: 'staticSPA',
        notify: true,
        minify: false,
        reloadDelay: 1000,
        browser: ['google chrome'],
        open: false
    };

    browserSync(options);
}

gulp.task('nodemon', false, function(cb) {
    var firstStart = true;
    var serverPort = 8000;
    $.nodemon({
        script: 'example/start.js',
        ext: 'js',
        env: {
            'NODE_ENV': 'development',
            'PORT': serverPort
        },
        nodeArgs: ['--debug'],
        ignore: [
            'coverage/**', 'node_modules/**',
            'gulpfile.js', '.idea/**', '.git/**'
        ],
        stdout: false // important for 'readable' event
    })
    // The http server might not have started listening yet when
    // the `restart` event has been triggered. It's best to check
    // whether it is already listening for connections or not.
    .on('readable', function() {
        this.stdout.on('data', function(chunk) {
            process.stdout.write(chunk);
            if (/listening at http/.test(chunk)) {
                startBrowserSync(serverPort);
                if (firstStart) {
                    firstStart = false;
                    cb();
                }
            }
        });
        this.stderr.pipe(process.stdout);
    });
    //.on('change', ['test-server'])
    //.on('start', function() {});
});

gulp.task('serve', 'Watch for file changes and re-run build and lint tasks', function(cb) {
    // When watch and nodemon tasks run at same time
    // the server seems to randomly blow up (??)
    runSequence(
        'watch',
        'nodemon',
        cb
    );
});
